### PR TITLE
Integrations: Upstage Layout Analysis Reader

### PR DESCRIPTION
**Pull request by junkeon**
# Description

Layout Analysis is one of Upstage's APIs, which detects document elements including tables and figures from documents. In this PR, we've introduced an Upstage reader based on the Layout Analysis API.

## New Package?

Did I fill in the `tool.llamahub` section in the `pyproject.toml` and provide a detailed README.md for my new integration or package?

- [x] Yes
- [ ] No

## Version Bump?

Did I bump the version in the `pyproject.toml` file of the package I am updating? (Except for the `llama-index-core` package)

- [x] Yes
- [ ] No

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Added new unit/integration tests
- [x] Added new notebook (that tests end-to-end)
- [ ] I stared at the code and made sure it makes sense

## Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added Google Colab support for the newly added notebooks.
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: llama-index-integrations/readers/llama-index-readers-upstage/README.md
Comment (logan-markewich) [inline llama-index-integrations/readers/llama-index-readers-upstage/README.md:1]: Great readme!

File: llama-index-integrations/readers/llama-index-readers-upstage/tests/examples/solar.pdf
Comment (logan-markewich) [inline llama-index-integrations/readers/llama-index-readers-upstage/tests/examples/solar.pdf:1]: tests that depend on external files require some extra steps to run properly in CICD with pantsbuild -- I don't have permissions to push the needed changes here.

File: llama-index-integrations/readers/llama-index-readers-upstage/poetry.lock
Comment (logan-markewich) [inline llama-index-integrations/readers/llama-index-readers-upstage/poetry.lock:1]: this can be removed

File: llama-index-integrations/readers/llama-index-readers-upstage/README.md
Comment (junkeon) [reply llama-index-integrations/readers/llama-index-readers-upstage/README.md:1]: Thank you for your kind comment

File: llama-index-integrations/readers/llama-index-readers-upstage/tests/examples/solar.pdf
Comment (junkeon) [reply llama-index-integrations/readers/llama-index-readers-upstage/tests/examples/solar.pdf:1]: Sorry for late. Would you tell me what should I do for solve this problem??

File: llama-index-integrations/readers/llama-index-readers-upstage/poetry.lock
Comment (junkeon) [reply llama-index-integrations/readers/llama-index-readers-upstage/poetry.lock:1]: OK. I will remove the file as your comment

File: llama-index-integrations/readers/llama-index-readers-upstage/tests/examples/solar.pdf
Comment (logan-markewich) [reply llama-index-integrations/readers/llama-index-readers-upstage/tests/examples/solar.pdf:1]: basically need to either give me permissions to push to your fork, or
- install [pants](https://www.pantsbuild.org/2.20/docs/getting-started/installing-pants)
- run `pants tailor llama-index-integrations/readers/::`
- add any BUILD files that it creates

File: llama-index-integrations/readers/llama-index-readers-upstage/README.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
+# UpstageLayoutAnalysisReader
+
+```bash
+pip install llama-index-readers-upstage
+```
+
+This reader loads document files and detects elements such as text, tables, and figures using the Upstage Layout Analysis API. Users wishing to utilize this reader must obtain an API access token from the [Upstage developers document](https://developers.upstage.ai/docs/getting-started/quick-start).
```
Comment (chosh0615) [inline llama-index-integrations/readers/llama-index-readers-upstage/README.md:7]:
```suggestion
This reader loads document files and detects elements such as text, tables, and figures using the Upstage Layout Analysis API. Users wishing to utilize this reader must obtain an API key from the [Upstage console](https://console.upstage.ai).
```

File: llama-index-integrations/readers/llama-index-readers-upstage/examples/upstage.ipynb [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    "Please install `llama_index` package.\n",
+    "```bash\n",
+    "pip install llama-index-readers-upstage\n",
+    "```\n",
+    "\n",
+    "## Environment Setup\n",
+    "Make sure to set the following environment variables:\n",
+    "- `UPSTAGE_API_KEY`: Your Upstage API key. Read [Upstage developers document](https://developers.upstage.ai/docs/getting-started/quick-start) to get your API key.\n",
+    "\n",
+    "    > The previously used UPSTAGE_DOCUMENT_AI_API_KEY is deprecated. However, the key previously used in UPSTAGE_DOCUMENT_AI_API_KEY can now be used in UPSTAGE_API_KEY.KJFG"
```
Comment (chosh0615) [inline llama-index-integrations/readers/llama-index-readers-upstage/examples/upstage.ipynb:21]: This line can be removed.

File: llama-index-integrations/readers/llama-index-readers-upstage/examples/upstage.ipynb [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    "\n",
+    "## Installation\n",
+    "Please install `llama_index` package.\n",
+    "```bash\n",
+    "pip install llama-index-readers-upstage\n",
+    "```\n",
+    "\n",
+    "## Environment Setup\n",
+    "Make sure to set the following environment variables:\n",
+    "- `UPSTAGE_API_KEY`: Your Upstage API key. Read [Upstage developers document](https://developers.upstage.ai/docs/getting-started/quick-start) to get your API key.\n",
```
Comment (chosh0615) [inline llama-index-integrations/readers/llama-index-readers-upstage/examples/upstage.ipynb:19]:
```suggestion
    "- `UPSTAGE_API_KEY`: Your Upstage API key. Read [Upstage console](https://console.upstage.ai/) to get your API key.\n",
```

File: llama-index-integrations/readers/llama-index-readers-upstage/llama_index/readers/upstage/base.py [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+            split (SplitType): The split type of the document.
+
+        Returns:
+            Document: A Document object representing the element with its content
+                      and metadata.
+
+        """
+        return Document(
+            text=(parse_output(element, output_type)),
+            extra_info={
```
Comment (chosh0615) [inline llama-index-integrations/readers/llama-index-readers-upstage/llama_index/readers/upstage/base.py:234]: Shouldn't we add bbox here?

Comment (junkeon) [reply llama-index-integrations/readers/llama-index-readers-upstage/llama_index/readers/upstage/base.py:234]: I have added the bounding boxes you mentioned.

File: llama-index-integrations/readers/llama-index-readers-upstage/examples/upstage.ipynb (hunk 2 above)
Comment (junkeon) [reply llama-index-integrations/readers/llama-index-readers-upstage/examples/upstage.ipynb:21]: I have removed the line.

File: llama-index-integrations/readers/llama-index-readers-upstage/tests/examples/solar.pdf
Comment (junkeon) [reply llama-index-integrations/readers/llama-index-readers-upstage/tests/examples/solar.pdf:1]: @logan-markewich 
I have incorporated the BUILD files based on your comments, thank you

**Comment by logan-markewich** [comment]
@junkeon ah, adding external files to tests is pretty annoying with pants, especially when I can't edit the PR 

Is there any way to avoid this? Can we just parse something that we can represent in code alone?

**Comment by chosh0615** [comment]
> @junkeon ah, adding external files to tests is pretty annoying with pants/CICD, especially when I can't edit the PR
> 
> Is there any way to avoid this? Can we just parse something that we can represent in code alone?

Thanks for checking the failure, @logan-markewich ! 
Like you said, we are facing some issues with the external file in unit test. Like other readers' test cases, we'll make another commit with these test cases removed for now!

File: .github/workflows/unit_test.yml [hunk 5]
```diff
@@ -11,7 +11,7 @@ env:
 
 jobs:
   test:
-    runs-on: ubuntu-latest-unit-tester
+    runs-on: ubuntu-latest
```
Comment (logan-markewich) [inline .github/workflows/unit_test.yml:14]: Oh, why was this changed? It will be 8x slower if we don't use the unit tester machine for running tests

**Review by logan-markewich — APPROVED** [review]

File: .github/workflows/unit_test.yml (hunk 5 above)
Comment (junkeon) [reply .github/workflows/unit_test.yml]: We reverted the changes made to the file for testing purposes in our repository.
